internal/validator/validator.go

```diff
@@ -0,0 +1,25 @@
+package validator
+
+type Validator struct {
+	Errors map[string]string
+}
+
+func New() *Validator {
+	return &Validator{Errors: make(map[string]string)}
+}
+
+func (v *Validator) IsValid() bool {
+	return len(v.Errors) == 0
+}
+
+func (v *Validator) AddError(key, message string) {
+	if _, exists := v.Errors[key]; !exists {
+		v.Errors[key] = message
+	}
+}
+
+func (v *Validator) Check(isValid bool, key, message string) {
+	if !isValid {
+		v.AddError(key, message)
+	}
+}
```

Comment:
Method naming is confusing, it is named as "check" and supposed to return check result like true/false, but it actually adds error to map. 
Why won't you just add error when it is really needed?
```
	if !validator.IsValidEmail(newEmail) {
		v.AddError("email", "invalid email")
		app.clientError(w, http.StatusUnprocessableEntity)
		return
	}
```

Reply:
Validator is supposed to run like this:

```
v := validator.New()
v.Check(validator.IsValidEmail(email), "email", "invalid email address")
v.Check(validator.Beetwen(age, 16, 45), "age", "age must be in rage 16-45")
v.Check(validator.Unique(values...), "value", "value must be unique")

if !v.IsValid() {
    app.Error()
}
```

The general approach is:

1. Create validator
2. Use built-in functions-checkers from validator package
3. In case you are not satisfied with built-in ones, just add error using `AddError()` method

Reply:
It's cool that you managed to extract validation logic into a separate package and incapsulate all related logic here 👍🏼

Experience of creating own validator seems useful.
However there are many validation libs in Go. JFYI guys: it's beneficial to not reinvent the wheel and use existed validator (e.g. [playground](https://github.com/go-playground/validator)) which already contains validation methods for many cases (email, ip address, etc)